File: c++/0042-trapping-rain-water/solution1.cpp
#include <algorithm>
#include <vector>

using namespace std;

class Solution1 {
public:
    int trap(vector<int> &height) {
        int size = height.size();

        vector<int> maxLeft(size);
        vector<int> maxRight(size);
        int result = 0;

        int currentMax = height[0];
        for (int i = 0; i < size; ++i) {
            maxLeft[i] = max(height[i], currentMax);
            currentMax = maxLeft[i];
        }

        currentMax = height[size - 1];
        for (int i = size - 1; i >= 0; --i) {
            maxRight[i] = max(height[i], currentMax);
            currentMax = maxRight[i];
        }

        for (int i = 0; i < size; ++i) {
            result += min(maxLeft[i], maxRight[i]) - height[i];
        }

        return result;
    }
};

File: java/0042-trapping-rain-water/Solution1.java
// O(n) memory solution
public class Solution1 {
    public int trap(int[] height) {
        int length = height.length;

        int maxLeft[] = new int[length];
        int maxRight[] = new int[length];
        int result = 0;

        int max = height[0];
        for (int i = 0; i < length; ++i) {
            maxLeft[i] = Math.max(height[i], max);
            max = maxLeft[i];
        }

        max = height[length - 1];
        for (int i = length - 1; i >= 0; --i) {
            maxRight[i] = Math.max(height[i], max);
            max = maxRight[i];
        }

        for (int i = 0; i < length; ++i) {
            result += Math.min(maxLeft[i], maxRight[i]) - height[i];
        }

        return result;
    }
}
